### PR TITLE
Migrated SIDH to KEM API (issue 350)

### PR DESCRIPTION
**Pull request by christianpaquin**
Migrated SIDH to KEM API, resolving issue #350.

### CODE REVIEW

**Comment by christianpaquin** [comment]
Fixed the issue on mac; @dstebila you can review and merge.

File: VisualStudio/oqs/oqs.vcxproj.filters [hunk 1]
```diff
@@ -7,9 +7,6 @@
     <ClCompile Include="..\..\src\common\rand.c">
       <Filter>common</Filter>
     </ClCompile>
-    <ClCompile Include="..\..\src\common\rand_nist.c">
-      <Filter>common</Filter>
-    </ClCompile>
```
Comment (dstebila) [inline VisualStudio/oqs/oqs.vcxproj.filters]: Why is this removed?

File: docs/algorithms/kem_sidh.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+| Parameter set   | Security model | Claimed NIST security level | Public key size (bytes) | Secret key size (bytes) | Ciphertext size (bytes) | Shared secret size (bytes) |
+|-----------------|:--------------:|:---------------------------:|:-----------------------:|:-----------------------:|:-----------------------:|:--------------------------:|
+| Sidh-p503       |     IND-CCA    |              1              |            378          |             32          |            378          |            126             |
+| Sidh-p751       |     IND-CCA    |              3              |            564          |             48          |            564          |            188             |
+
+Implementation
+--------------
+
+- **Source of implementation:** https://github.com/Microsoft/PQCrypto-SIDH
```
Comment (dstebila) [inline docs/algorithms/kem_sidh.md]: Can you put a version number, date, or commit hash of the source implementation?  This will make it easier to know where to look when updating implementations later.

File: docs/algorithms/kem_sidh.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- **License:** MIT License
+- **Language:** C
+- **Constant-time:** Yes
+- **Architectures supported in liboqs master branch**: x86, x64
+
+Additional comments
+-------------------
+
+
```
Comment (dstebila) [inline docs/algorithms/kem_sidh.md]: Maybe add a comment that KAT checks are not used for this algorithm.

File: src/kem/sike/Makefile.am [hunk 4]
```diff
@@ -3,7 +3,7 @@ noinst_LTLIBRARIES = libkemsike.la
 
 libkemsike_la_SOURCES = kem_sike.c P503/P503.c P751/P751.c
 
-libkemsike_la_CFLAGS = $(AM_CFLAGS) -w
+libkemsike_la_CFLAGS = $(AM_CFLAGS) -w -g
```
Comment (dstebila) [inline src/kem/sike/Makefile.am]: Do you really want debugging symbols here?

File: src/kem/sike/kem_sike.c [hunk 5]
```diff
@@ -70,3 +70,159 @@ OQS_KEM *OQS_KEM_sike_p751_new() {
...
+	kem->keypair = OQS_KEM_sidh_p503_keypair;
+	kem->encaps = OQS_KEM_sidh_p503_encaps;
+	kem->decaps = OQS_KEM_sidh_p503_decaps;
+
+	return kem;
+}
+
+/* Forward KEM API calls to SIDH's API */
+OQS_STATUS OQS_KEM_sidh_p503_keypair(uint8_t *public_key, uint8_t *secret_key) {
+  if (oqs_kem_sidh_p503_EphemeralKeyGeneration_A((unsigned char *) secret_key, (unsigned char *) public_key) != 0) {
```
Comment (dstebila) [inline src/kem/sike/kem_sike.c]: Tabs not spaces

File: src/kem/sike/kem_sike.c [hunk 6]
```diff
@@ -70,3 +70,159 @@ OQS_KEM *OQS_KEM_sike_p751_new() {
 }
 
 #endif
+
+// Preventing redefinition of symbols
+// Todo: remove
+#undef OQS_SIDH_MSR_CRYPTO_SECRETKEYBYTES
```
Comment (dstebila) [inline src/kem/sike/kem_sike.c]: You've marked this `Todo: remove` -- is it to be removed before merging or something we should do later (which we are likely to forget)?

File: src/kem/sike/Makefile.am (hunk 4 above)
Comment (christianpaquin) [reply src/kem/sike/Makefile.am]: Oups, thanks for catching that; leftovers from a debugging session

File: src/kem/sike/kem_sike.c [hunk 7]
```diff
@@ -70,3 +70,159 @@ OQS_KEM *OQS_KEM_sike_p751_new() {
...
+OQS_KEM *OQS_KEM_sidh_p503_new() {
+
+	OQS_KEM *kem = malloc(sizeof(OQS_KEM));
+	if (kem == NULL) {
+		return NULL;
+	}
+	kem->method_name = OQS_KEM_alg_sidh_p503;
+
+	kem->claimed_nist_level = 1;
+	kem->ind_cca = true;
```
Comment (dstebila) [inline src/kem/sike/kem_sike.c]: SIDH is not IND-CCA-secure, if I understand correctly.

File: src/kem/sike/kem_sike.c [hunk 8]
```diff
@@ -70,3 +70,159 @@ OQS_KEM *OQS_KEM_sike_p751_new() {
...
+OQS_KEM *OQS_KEM_sidh_p751_new() {
+
+	OQS_KEM *kem = malloc(sizeof(OQS_KEM));
+	if (kem == NULL) {
+		return NULL;
+	}
+	kem->method_name = OQS_KEM_alg_sidh_p751;
+
+	kem->claimed_nist_level = 3;
+	kem->ind_cca = true;
```
Comment (dstebila) [inline src/kem/sike/kem_sike.c]: SIDH is not IND-CCA-secure, if I understand correctly.

File: src/kem/sike/kem_sike.c [hunk 9]
```diff
@@ -70,3 +70,159 @@ OQS_KEM *OQS_KEM_sike_p751_new() {
...
+#undef OQS_SIDH_MSR_CRYPTO_BYTES
+#undef OQS_SIDH_MSR_CRYPTO_CIPHERTEXTBYTES
+#undef OQS_SIDH_MSR_CRYPTO_ALGNAME
+#undef SIDH_SECRETKEYBYTES
+#undef SIDH_PUBLICKEYBYTES
+#undef SIDH_BYTES
+
+#ifdef OQS_ENABLE_KEM_sidh_p503
+
+#include "P503/P503_api.h"
```
Comment (dstebila) [inline src/kem/sike/kem_sike.c]: Why do you need to include P503\_api.h?

Comment (dstebila) [reply src/kem/sike/kem_sike.c]: Nevermind, I see now.

File: VisualStudio/oqs/oqs.vcxproj.filters (hunk 1 above)
Comment (christianpaquin) [reply VisualStudio/oqs/oqs.vcxproj.filters]: This was added in PR #348, and my PR was prepared before that landed. I'll merge with the latest.

File: docs/algorithms/kem_sidh.md (hunk 2 above)
Comment (christianpaquin) [reply docs/algorithms/kem_sidh.md]: Yes.

File: docs/algorithms/kem_sidh.md (hunk 3 above)
Comment (christianpaquin) [reply docs/algorithms/kem_sidh.md]: Good idea.

**Comment by christianpaquin** [comment]
Thanks for your review @dstebila; changes pushed.

**Comment by christianpaquin** [comment]
Simultaneously merged #358 now conflicts with this. I'm cleaning it up...

**Comment by christianpaquin** [comment]
Merge conflicts resolved; @dstebila ready for merge.
